### PR TITLE
DEV: Rearchitect chat tracking state

### PR DESCRIPTION
**Pull request by martin-brennan**
This moves chat tracking state calculation for channels
and threads into a central Chat::TrackingStateManager service, that
serves a similar purpose to the TopicTrackingState model
in core.

This service calls down to these query classes:

* ThreadUnreadsQuery
* ChannelUnreadsQuery

To get the unread_count and mention_count for the appropriate
channels and threads.

As well as this, this commit refactors the client-side chat
tracking state.

Now, there is a central ChatTrackingStateManager Ember Service
so all tracking is accessible and can be counted from one place,
which can also initialize tracking from an initial payload.

The actual tracking counts are now maintained in a ChatTrackingState
class that is initialized on the `.tracking` property of both channel and
thread objects.

This removes the attributes on UserChatChannelMembership and decoration
of said membership from ChannelFetcher, preferring instead to have an additional
object for tracking in the JSON.

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/initializers/chat-setup.js [hunk 1]
```diff
@@ -122,6 +122,9 @@ export default {
       document.body.classList.add("chat-enabled");
 
       const currentUser = api.getCurrentUser();
+
+      // NOTE: chat_channels is more than a simple array, it also contains
+      // tracking and membership data, see Chat::StructuredChannelSerializer
       if (currentUser?.chat_channels) {
```
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/initializers/chat-setup.js:128]: Maybe we want to rename `chat_channels` at some point...not sure what to though

File: plugins/chat/assets/javascripts/discourse/services/chat-tracking-state-manager.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  @tracked _mentionCount = 0;
+
+  constructor(owner, params = {}) {
+    setOwner(this, owner);
+    this._unreadCount = params.hasOwnProperty("unreadCount")
+      ? params.unreadCount
+      : 0;
+    this._mentionCount = params.hasOwnProperty("mentionCount")
+      ? params.mentionCount
+      : 0;
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/services/chat-tracking-state-manager.js:19]: `params.unreadCount || 0`
`params.mentionCount || 0`

File: plugins/chat/assets/javascripts/discourse/services/chat-tracking-state-manager.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    this._unreadCount = value;
+  }
+
+  get mentionCount() {
+    return this._mentionCount;
+  }
+
+  set mentionCount(value) {
+    const valueChanged = this._mentionCount !== value;
+    this._mentionCount = value;
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/services/chat-tracking-state-manager.js:41]: Maybe we could move this in if valueChanged?

File: plugins/chat/assets/javascripts/discourse/services/chat-tracking-state-manager.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  #setState(model, state) {
+    if (!model.tracking) {
+      model.tracking = new ChatTrackingState(getOwner(this), {
+        unreadCount: state.unread_count,
+        mentionCount: state.mention_count,
+      });
+      return;
+    }
+    model.tracking.unreadCount = state.unread_count;
+    model.tracking.mentionCount = state.mention_count;
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/services/chat-tracking-state-manager.js:125]:
```javascript
if (!model.tracking) {
  model.tracking = new ChatTrackingState(getOwner(this),);
}
model.tracking.unreadCount = state.unread_count;
model.tracking.mentionCount = state.mention_count;
```

**Review by jjaffeux — APPROVED** [review]

**Review by jjaffeux — APPROVED** [review]

**Pull request merged**